File: ios/RCTTwilioIPMessaging/RCTTwilioIPMessagingMembers.m
//
//  RCTTwilioIPMessagingMembers.m
//  TwilioIPExample
//
//  Created by Brad Bumbalough on 6/7/16.
//  Copyright © 2016 Facebook. All rights reserved.
//

#import "RCTTwilioIPMessagingMembers.h"
#import "RCTConvert+TwilioIPMessagingClient.h"
#import "RCTTwilioIPMessagingClient.h"
#import <RCTUtils.h>

@implementation RCTTwilioIPMessagingMembers

RCT_EXPORT_MODULE()

- (TWMMembers *)loadMembersFromChannelSid:(NSString *)sid {
  TwilioIPMessagingClient *client = [[RCTTwilioIPMessagingClient sharedManager] client];
  return [[[client channelsList] channelWithId:sid] members];
}

#pragma mark Members Methods

RCT_REMAP_METHOD(getMembers, channelSid:(NSString *)channelSid allObjects_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSArray<TWMMember *> *members = [[self loadMembersFromChannelSid:channelSid] allObjects];
  NSMutableArray *response = [NSMutableArray array];
  if (members) {
    for (TWMMember *member in members) {
      [response addObject:[RCTConvert TWMMember:member]];
    }
  }
  resolve(RCTNullIfNil(response));
}

RCT_REMAP_METHOD(add, channelSid:(NSString *)channelSid identity:(NSString *)identity add_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  [[self loadMembersFromChannelSid:channelSid] addByIdentity:identity completion:^(TWMResult *result) {
    if (result.isSuccessful) {
      resolve(@[@TRUE]);
    }
    else {
      reject(@"add-identity-error", @"error", result.error);
    }
  }];
}

RCT_REMAP_METHOD(invite, channelSid:(NSString *)channelSid identity:(NSString *)identity invite_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  [[self loadMembersFromChannelSid:channelSid] inviteByIdentity:identity completion:^(TWMResult *result) {
    if (result.isSuccessful) {
      resolve(@[@TRUE]);
    }
    else {
      reject(@"invite-identity-error", @"error", result.error);
    }
  }];
}

RCT_REMAP_METHOD(remove, channelSid:(NSString *)channelSid identity:(NSString *)identity remove_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  TwilioIPMessagingClient *client = [[RCTTwilioIPMessagingClient sharedManager] client];
  TWMMember *member = [[[client channelsList] channelWithId:channelSid] memberWithIdentity:identity];
  [[self loadMembersFromChannelSid:channelSid] removeMember:member completion:^(TWMResult *result) {
    if (result.isSuccessful) {
      resolve(@[@TRUE]);
    }
    else {
      reject(@"remove-member-error", @"error", result.error);
    }
  }];
}


RCT_REMAP_METHOD(getMember, channelSid:(NSString *)channelSid identity:(NSString *)identity member_resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    resolve([RCTConvert TWMMember:[[self loadMembersFromChannelSid:channelSid] memberWithIdentity:identity]);
}

@end

File: android/app/src/main/java/com/bradbumbalough/twilioipmessaging/RCTTwilioIPMessagingMembers.java
package com.bradbumbalough.twilioipmessaging;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;

import com.twilio.ipmessaging.Member;
import com.twilio.ipmessaging.Constants;
import com.twilio.ipmessaging.ErrorInfo;
import com.twilio.ipmessaging.Members;

public class RCTTwilioIPMessagingMembers extends ReactContextBaseJavaModule {

    @Override
    public String getName() {
        return "TwilioIPMessagingMembers";
    }

    private ReactApplicationContext reactContext;

    public RCTTwilioIPMessagingMembers(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    private Members loadMembersFromChannelSid(String sid) {
        return RCTTwilioIPMessagingClient.getInstance().client.getChannels().getChannel(sid).getMembers();
    }

    @ReactMethod
    public void getMembers(String channelSid, Promise promise) {
        promise.resolve(loadMembersFromChannelSid(channelSid).getMembers());
    }

    @ReactMethod
    public void add(String channelSid, String identity, final Promise promise) {

        Constants.StatusListener listener = new Constants.StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
            }

            @Override
            public void onSuccess() {
                promise.resolve(true);
            }
        };
        loadMembersFromChannelSid(channelSid).addByIdentity(identity, listener);
    }

    @ReactMethod
    public void invite(String channelSid, String identity, final Promise promise) {

        Constants.StatusListener listener = new Constants.StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
            }

            @Override
            public void onSuccess() {
                promise.resolve(true);
            }
        };
        loadMembersFromChannelSid(channelSid).inviteByIdentity(identity, listener);
    }

    @ReactMethod
    public void remove(String channelSid, String identity, final Promise promise) {

        Constants.StatusListener listener = new Constants.StatusListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                super.onError(errorInfo);
            }

            @Override
            public void onSuccess() {
                promise.resolve(true);
            }
        };
        Member[] members = loadMembersFromChannelSid(channelSid).getMembers();
        Member memberToDelete = null;

        for (Member m : members) {
            if (m.getUserInfo().getIdentity() == identity) {
                memberToDelete = m;
                break;
            }
        }
        loadMembersFromChannelSid(channelSid).removeMember(memberToDelete, listener);
    }

    @ReactMethod
    public void getMember(String channelSid, String identity, final Promise promise) {
        Member[] members = loadMembersFromChannelSid(channelSid).getMembers();

        for (Member m : members) {
            if (m.getUserInfo().getIdentity() == identity) {
                promise.resolve(m);
                break;
            }
        }
    }

}
